### PR TITLE
feat : Move `IsUniform`, refactor

### PR DESCRIPTION
**Pull request by JADekker**
Move `IsUniform` from `Probability/Density` to
`Probability/Distributions/Uniform` to make it more in line with the other distributions.

Refactor `IsUniform` in terms of the new `uniformVolume`, which abstracts away from the random variable and probability space. 

This PR will attempt to also refactor
`Probability/ProbabilityMassFunction/Uniform` in such a way that this follows naturally from `IsUniform`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Probability/Distributions/Uniform.lean [hunk 1]
```diff
@@ -0,0 +1,217 @@
+/-
+Copyright (c) 2024 Josha Dekker. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Josha Dekker
```
Comment (RemyDegenne) [inline Mathlib/Probability/Distributions/Uniform.lean:4]:
```suggestion
Authors: Josha Dekker, Kexing Ying
```

File: Mathlib/Probability/Distributions/Uniform.lean [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+#align measure_theory.pdf.is_uniform.integral_eq MeasureTheory.pdf.IsUniform.integral_eq
+
+end IsUniform
+
+lemma uniformVolume.IsUniform {s : Set E} : IsUniform (id : E → E) s (uniformVolume s μ) μ := by
+  unfold IsUniform
+  rw [map_id]
+
+/-- Measure defined by a uniform distribution on s. -/
+def uniformMeasure (s : Set E) (μ : Measure E := by volume_tac) : Measure E := uniformVolume s μ
```
Comment (RemyDegenne) [inline Mathlib/Probability/Distributions/Uniform.lean:212]: This is an exact duplicate of `uniformVolume`? You should remove one of them. I think `uniformMeasure` is the better name (I don't understand the "volume" part of the name uniformVolume).

File: Mathlib/Probability/Distributions/Uniform.lean [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+    uniformVolume s μ ≪ μ := by
+  intro t ht
+  unfold uniformVolume
+  rw [smul_apply, smul_eq_mul]
+  apply mul_eq_zero.mpr
+  refine Or.inr (le_antisymm ?_ (zero_le _))
+  exact ht ▸ restrict_apply_le s t
+
+theorem measure_preimage {s : Set E} {A : Set E}
+    (hA : MeasurableSet A) (hunif : m = uniformVolume s μ): m A = μ (s ∩ A) / μ s := by
```
Comment (RemyDegenne) [inline Mathlib/Probability/Distributions/Uniform.lean:47]: `m` is useless here. Simply replace it with `uniformVolume s μ` in the statement. Also I don't understand the name: there is no preimage involved.

You could remove that lemma from the namespace UniformVolume (and consider removing all lemmas from that namespace) and rename it `uniformVolume_apply`.

**Review by RemyDegenne — COMMENTED** [review]
I know this is still WIP, but here are some comments nonetheless.

File: Mathlib/Probability/Distributions/Uniform.lean (hunk 3 above)
Comment (JADekker) [reply Mathlib/Probability/Distributions/Uniform.lean:47]: Thank you for the feedback! I'm not sure what you mean by "removing all lemmas from the namespace UniformVolume": you mean that I should get rid of the lemmas, or just move them up one level?

File: Mathlib/Probability/Distributions/Uniform.lean (hunk 2 above)
Comment (JADekker) [reply Mathlib/Probability/Distributions/Uniform.lean:212]: The reason I went for `uniformVolume` was that it requires a volume measure (or at least, most things go through for general measures, but the crucial fact that our PDF becomes an indicator function hinges on this assumption). But let me rewrite it in terms of uniformMeasure, that is certainly more elegant.

Comment (JADekker) [reply Mathlib/Probability/Distributions/Uniform.lean]: Apparently that is not really the case, I've cleaned it up.

File: Mathlib/Probability/Distributions/Uniform.lean [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+
+namespace pdf
+
+variable {Ω : Type*}
+
+variable {_ : MeasurableSpace Ω} {ℙ : Measure Ω}
+
+/-- A random variable `X` has uniform distribution on `s` if its push-forward measure is
+`(μ s)⁻¹ • μ.restrict s`. -/
+def IsUniform (X : Ω → E) (s : Set E) (ℙ : Measure Ω) (μ : Measure E) :=
```
Comment (RemyDegenne) [inline Mathlib/Probability/Distributions/Uniform.lean:76]: Why did you remove `by volume_tac`? The role of `by volume_tac` is to automatically fill the measure with `volume`, in case the space is actually a measure space and a volume is available (which is the main case we care about). It does not prevent you from using the def with other measures.

Comment (JADekker) [reply Mathlib/Probability/Distributions/Uniform.lean:76]: Because I misunderstood by volume_tac, my bad. I'll put it back!

**Comment by JADekker** [comment]
Just to update on this: I'm currently too busy, but we basically need one more step to integrate the `uniform` stuff from ProbabilityMassFunction in a "uniform" way here.... I'll copy my message from the relevant commit here: 

The next step would be to show that our usual PDF agrees with the one
that we picked, and then to derive the PMF from there. I have added
two definitions and show that they agree with `pdf ...` and each-other.

I need to link this to PMF, but I'm stuck on getting this to work.
I probably won't have time to work on this the next days, so feel free
to pick this up.

**Comment by Ruben-VandeVelde** [comment]
I recommend against changing code in the same PR as moving it into a new file; I imagine a preliminary PR that only creates the new file would be welcome, though.

**Comment by JADekker** [comment]
I’ll make such a PR in a bit!
